### PR TITLE
[Security] Migrate the session on login only when the user changes

### LINKED ISSUE #49194: "The CSRF token is invalid." in tests after upgrading to 6.2.6 to fix CVE
### Symfony version(s) affected

6.2.6

### Description

After installing the fix for https://symfony.com/blog/cve-2022-24895-csrf-token-fixation, tests fail with the message "The CSRF token is invalid." I have only been able to reproduce this in tests, not (yet) when running the application in the browser.

@nicolas-grekas

### How to reproduce

1. Install the Symfony demo project: https://github.com/symfony/demo
2. Run `./bin/phpunit`. All tests should succeed.
3. Run `composer update`.
4. Run `./bin/phpunit`. Multiple tests should fail. Add `| grep CSRF` and you will find the error message.

### Possible Solution

I have no clue yet. My conclusion is that `$this->csrfTokenStorage->clear();` is the cause of this bug (see https://github.com/symfony/symfony/commit/5909d74ecee359ea4982fcf4331aaf2e489a1fd4#diff-0ff1412624a79146c346925f2407eb4783b144da38ddb369ca30e49d046fab70R59), but removing this is obviously not an option as it is the fix for the CVE.

### Additional Context

Docker one-liner:

```shell
docker run -it --rm composer bash -c " \
  composer create-project symfony/symfony-demo my_project; \
  cd my_project/; \
  ./bin/phpunit; \
  composer update; \
  ./bin/phpunit | head -n 6; \
  ./bin/phpunit | grep CSRF \
"

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #49194
| License       | MIT
| Doc PR        | -

As described in the linked issue, the recent security fix breaks submitting forms when stateless authenticators are mixed with stateful CSRF storage.

This PR fixes the issue by ensuring that the session is *not* migrated when the user selected by the stateless authenticator is the same as the one retrieved from the session.

In order to do so, I had add to the previous token to `LoginSuccessEvent`.

/cc @wouterj @weaverryan @stof can you please have a look? This should be part of the next release if possible.

### CODE REVIEW

File: src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php [hunk 1]
```diff
@@ -37,14 +38,15 @@ class LoginSuccessEvent extends Event
     private $authenticator;
     private $passport;
     private $authenticatedToken;
+    private $previousToken;
```
Comment (stof) [inline src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php:40]: should we use a typed property ?

File: src/Symfony/Component/Security/Http/EventListener/SessionStrategyListener.php [hunk 2]
```diff
@@ -39,7 +39,7 @@ public function onSuccessfulLogin(LoginSuccessEvent $event): void
         $request = $event->getRequest();
         $token = $event->getAuthenticatedToken();
 
-        if (!$request->hasSession() || !$request->hasPreviousSession()) {
+        if (!$request->hasSession() || !$request->hasPreviousSession() || $event->getPreviousToken()->getUser() === ($token->getUser() ?? false)) {
```
Comment (stof) [inline src/Symfony/Component/Security/Http/EventListener/SessionStrategyListener.php]: should we check the identity of the user instance or should we check the user identifier ?

A previous token might contain an unserialized user, which could be a different PHP object instance than the object returned by the user provider, while still being the same authenticated user from a business point of view.

File: src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php:40]: Not on the 5.4 branch.

Comment (stof) [reply src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php:40]: indeed

File: src/Symfony/Component/Security/Http/EventListener/SessionStrategyListener.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Http/EventListener/SessionStrategyListener.php]: At this point, the token has been refreshed by ContextListener. The situation you describe might exist though, I'm no expert here. My proposal it to start this way and see later if we need more fuzzy matching (what ContextListener already implements.)

**Review by chalasr — APPROVED** [review]

**Review by wouterj — CHANGES_REQUESTED** [review]
I've shared this in Slack, but to make sure it's visible: I don't believe this fix fixes anything, except from edge cases you get when using the HTTP basic trick in tests.

I've created a reproducer to test HTTP basic on both stateless and stateful firewalls together with CSRF in a browser: https://github.com/wouterj/sf-reproducer/tree/issue-49526

It seems like both cases still give the error:
* There is no previous token on stateless firewalls, which means the new check always fails.
* On stateful firewalls, there is a token, but it holds a different instance of the user, which also make the new check fail.

**Comment by wouterj** [comment]
Aside from my previous comment, I also wonder if the legacy security system isn't affected by this? If it's not, it would be good to compare as I don't think we explicitly changed this behavior. If it is, we should also provide a fix for it in the 5.4 branch :(

**Comment by nicolas-grekas** [comment]
Thanks for the feedback. I updated the implementation to compare using the user identifier (or username) and I also updated the legacy system.

Status: needs review

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php [hunk 3]
```diff
@@ -56,9 +56,9 @@ public function __construct(TokenStorageInterface $tokenStorage, EventDispatcher
     /**
      * Authenticates the given token in the system.
      */
-    public function authenticateWithToken(TokenInterface $token, Request $request, string $providerKey = null)
+    public function authenticateWithToken(TokenInterface $token, Request $request, string $providerKey = null, TokenInterface $previousToken = null)
     {
-        $this->migrateSession($request, $token, $providerKey);
+        $this->migrateSession($request, $token, $providerKey, 3 < \func_num_args() ? $previousToken : $this->tokenStorage->getToken());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php:61]: This class is `@final` so adding the arg is not a BC break.

For extra safety, I ensured the previous token should always be read *before* calling `authenticate()`. In case the caller hasn't been updated to the new signature, this check falls back to reading the previous token here.

Comment (wouterj) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php:61]: Is there something against not adding this argument and doing `$this->tokenStorage->getToken()` in `migrateSession()`? That seems a bit safer BC wise, given this is an LTS release.

File: src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php [hunk 4]
```diff
@@ -53,6 +55,7 @@ public function __construct(AuthenticatorInterface $authenticator, PassportInter
         $this->authenticator = $authenticator;
         $this->passport = $passport;
         $this->authenticatedToken = $authenticatedToken;
+        $this->previousToken = $previousToken ?? new NullToken();
```
Comment (wouterj) [inline src/Symfony/Component/Security/Http/Event/LoginSuccessEvent.php]: I would like to not use `NullToken` anywhere outside the authorization checker, as they are a major exception to the "tokens are always authenticated" rule in the new security system.

Let's make `getPreviousToken(): ?TokenInterface` instead.

File: src/Symfony/Component/Security/Http/Tests/EventListener/SessionStrategyListenerTest.php [hunk 5]
```diff
@@ -62,6 +62,25 @@ public function testStatelessFirewalls()
         $listener->onSuccessfulLogin($this->createEvent('api_firewall'));
     }
 
+    public function testRequestWithoutSamePreviousUser()
```
Comment (wouterj) [inline src/Symfony/Component/Security/Http/Tests/EventListener/SessionStrategyListenerTest.php]:
```suggestion
    public function testRequestWithSamePreviousUser()
```

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php:61]: `$this->tokenStorage->getToken()` should be called before calling `authenticate()`, that's why.
I still want a fallback path because getting a proper previous token requires upgrading both security-bundle (to wire the new argument)

**Review by wouterj — APPROVED** [review]
Ready to merge imho.

@chalasr would you mind having a second look, given a lot more changes were introduced (in the old security system) since your approval?
If it's all ok, I can do the merge & up-merge of this one.

**Review by chalasr — APPROVED** [review]
Carefully reviewed, looks good to me. Thanks

**Comment by wouterj** [comment]
Thanks for the work @nicolas-grekas and reviews @chalasr :)

**Comment by gnutix** [comment]
Shouldn't this be released for all versions where the security fix that causes the issue was released (like 6.0.20) ? (I'm upgrading to 6.2 anyway, but maybe others can't do it right now...)

**Comment by wouterj** [comment]
You can always safely upgrade to 6.2 from 6.0 or 6.1, there are no BC breaks in minor versions.

Those versions have reached end of life last month and will not receive new fixes.
